src/three-components/Shadow.ts

```diff
@@ -13,26 +13,103 @@
- */
-export default class Shadow extends Mesh {
-  /**
-   * Create a shadow mesh.
-   */
-  constructor() {
-    const geometry = new PlaneGeometry(2000, 2000);
-    geometry.rotateX(-Math.PI / 2);
-
-    const material = new ShadowMaterial(({
-      color: 0x111111,
-      opacity: 0.2,
-    } as any));
-
-    super(geometry, material);
-    this.receiveShadow = true;
-    this.name = 'ShadowMesh';
+import ModelScene from './ModelScene';
+
+// Nothing within shadowOffset of the bottom of the model casts a shadow
+// (this is to avoid having a baked-in shadow plane cast its own shadow).
+const SHADOW_OFFSET = 0.001;
+const BASE_SHADOW_OPACITY = 0.1;
+// The softness of the shadow is controlled with this resolution parameter. The
+// lower the resolution, the softer the shadow.
+const MAX_SHADOW_RESOLUTION = 64;
+
+export class Shadow extends DirectionalLight {
+  private shadowMaterial = new ShadowMaterial;
+  private floor: Mesh;
+  public needsUpdate = false;
+
+  constructor(private scene: ModelScene) {
+    super();
+
+    // We use the light only to cast a shadow, not to light the scene.
+    this.intensity = 0;
+    this.castShadow = true;
+
+    this.floor = new Mesh(new PlaneBufferGeometry, this.shadowMaterial);
+    this.floor.receiveShadow = true;
+    this.floor.castShadow = false;
+    scene.pivot.add(this.floor);
+    scene.pivot.add(this);
+    this.target = scene.pivot;
+
+    this.setScene(scene);
+  }
+
+  setScene(scene: ModelScene) {
+    this.scene = scene;
+    const {boundingBox, size} = scene.model;
+    const {camera} = this.shadow;
+
+    const shadowOffset = size.y * SHADOW_OFFSET;
+
+    this.floor.rotateX(-Math.PI / 2);
+    boundingBox.getCenter(this.floor.position);
+    // Floor plane is up slightly to avoid Z-fighting with baked-in shadows and
+    // to stay inside the shadow camera.
+    this.floor.position.y -= size.y / 2 - 2 * shadowOffset;
+
+    this.position.y = boundingBox.max.y + shadowOffset;
+    this.up.set(0, 0, 1);
+    camera.near = 0;
+    camera.far = size.y;
+
+    this.setMapSize(MAX_SHADOW_RESOLUTION);
+  }
+
+  setMapSize(maxMapSize: number) {
+    const {boundingBox, size} = this.scene.model;
+    const {camera, mapSize} = this.shadow;
+    const width =
+        size.x > size.z ? maxMapSize : Math.floor(maxMapSize * size.x / size.z);
+    const height =
+        size.x > size.z ? Math.floor(maxMapSize * size.z / size.x) : maxMapSize;
+
+    mapSize.set(width, height);
+    // These pads account for the softening radius around the shadow.
+    const widthPad = 2.5 * size.x / width;
+    const heightPad = 2.5 * size.z / height;
+
+    camera.left = -boundingBox.max.x - widthPad;
+    camera.right = -boundingBox.min.x + widthPad;
+    camera.bottom = boundingBox.min.z - heightPad;
+    camera.top = boundingBox.max.z + heightPad;
+
+    this.updateMatrixWorld();
+    (this.shadow as any).updateMatrices(this);
+
+    this.floor.scale.set(size.x + 2 * widthPad, size.z + 2 * heightPad, 1);
```

Comment:
Yeah, it seems like the easiest near-term thing to do would be to just scale the floor up by a multiple of the bounding-box footprint. Call it 3x? Animated models that move around much more than that are gonna look very poorly framed anyway...

And in the future, maybe we can explore clever improvements like re-using our theorized filmstrip method for fidelity screenshots to compute an approximation of an animation's effective bounding space.

Reply:
I went for 2x, and made it a cube based on the largest dimension. Figured that's better since many humanoid models rest with arms out, so their bounding boxes are pretty high aspect ratio.